lib/smtp.c

```diff
@@ -542,8 +542,25 @@ static CURLcode smtp_perform_mail(struct connectdata *conn)
 
   /* Calculate the optional AUTH parameter */
   if(data->set.str[STRING_MAIL_AUTH] && conn->proto.smtpc.sasl.authused) {
-    if(data->set.str[STRING_MAIL_AUTH][0] != '\0')
-      auth = aprintf("%s", data->set.str[STRING_MAIL_AUTH]);
+    if(data->set.str[STRING_MAIL_AUTH][0] != '\0') {
+      struct hostname mail_auth = { NULL, NULL, NULL, NULL };
+
+      mail_auth.name = data->set.str[STRING_MAIL_AUTH];
+
+      /* Convert the MAIL AUTH address to IDN (if supported by the server) */
+      if(conn->proto.smtpc.utf8_supported) {
+        result = Curl_idnconvert_hostname(conn, &mail_auth);
+        if(result) {
+          free(from);
+
+          return result;
+        }
+      }
+
+      auth = aprintf("%s", mail_auth.name);
```

Comment:
If this returns NULL we need to set `result` too for failure, right?

Reply:
Yep - isn't that take care of by the block that starts at line 568? Unless I've screwed up the indentation:
````
    if(!auth) {
      free(from);

      return CURLE_OUT_OF_MEMORY;
    }
`````